apitools/base/py/transfer_test.py

```diff
@@ -12,6 +14,146 @@
+        for start in (0, 10):
+            self.assertEqual(total_size - 1,
+                             download._Download__ComputeEndByte(start),
+                             msg='Failed on start={0}'.format(start))
+
+    def testNonChunkedDownload(self):
+        bytes_http = object()
+        http = object()
+        download_stream = six.StringIO()
+        download = transfer.Download.FromStream(download_stream, total_size=52)
+        download.bytes_http = bytes_http
+        base_url = 'https://part.one/'
+
+        with mock.patch.object(http_wrapper, 'MakeRequest',
+                               autospec=True) as make_request:
+            make_request.return_value = http_wrapper.Response(
+                info={
+                    'content-range': 'bytes 0-51/52',
+                    'status': http_client.OK,
+                },
+                content=string.ascii_lowercase * 2,
+                request_url=base_url,
+            )
+            request = http_wrapper.Request(url='https://part.one/')
+            download.InitializeDownload(request, http=http)
+            self.assertEqual(1, make_request.call_count)
+            received_request = make_request.call_args[0][1]
+            self.assertEqual(base_url, received_request.url)
+            self.assertRangeAndContentRangeCompatible(
+                received_request, make_request.return_value)
+            download_stream.seek(0)
+            self.assertEqual(string.ascii_lowercase * 2,
+                             download_stream.getvalue())
+
+    def testChunkedDownload(self):
+        bytes_http = object()
+        http = object()
+        download_stream = six.StringIO()
+        download = transfer.Download.FromStream(
+            download_stream, chunksize=26, total_size=52)
+        download.bytes_http = bytes_http
+
+        # Setting autospec on a mock with an iterable side_effect is
+        # currently broken (http://bugs.python.org/issue17826), so
+        # instead we write a little function.
+        def _ReturnBytes(unused_http, http_request,
+                         *unused_args, **unused_kwds):
+            url = http_request.url
+            if url == 'https://part.one/':
+                return http_wrapper.Response(
+                    info={
+                        'content-location': 'https://part.two/',
+                        'content-range': 'bytes 0-25/52',
+                        'status': http_client.PARTIAL_CONTENT,
+                    },
+                    content=string.ascii_lowercase,
+                    request_url='https://part.one/',
+                )
+            elif url == 'https://part.two/':
+                return http_wrapper.Response(
+                    info={
+                        'content-range': 'bytes 26-51/52',
+                        'status': http_client.OK,
+                    },
+                    content=string.ascii_lowercase,
```

Comment:
It would probably be good to differentiate the bytes in this case (so that the second response is different in content than the first)